ArtOfIllusion/src/artofillusion/PluginRegistry.java

```diff
@@ -36,6 +35,35 @@ Changes copyright (C) 2018-2022 by Maksim Khramov
   private static final HashMap<String, ExportInfo> exports = new HashMap<String, ExportInfo>();
   private static final HashMap<String, Object> classMap = new HashMap<String, Object>();
 
+  static {
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+        List<PluginRegistry.PluginThrowable> errors = PluginRegistry.notifyPlugins(Plugin.class, Plugin.APPLICATION_STOPPING);
+        errors.forEach(error -> System.out.println("Plugin " + error.getSource().getClass().getSimpleName() + " throws error: " + error.getCause().getMessage()) );
+    }, "Plugin shutdown thread"));
+  }
+  
+  
+  public static <T> List<PluginThrowable> notifyPlugins(Class<T> category, int message, Object... args)
+  {
+    if(categoryClasses.containsKey(category))
+    {
+        List<PluginThrowable> errors = new ArrayList<>();
+        for(Object plugin: categoryClasses.get(category))
+        {
+          try 
+          {
+            ((Plugin)plugin).processMessage(message, args);
+          } 
+          catch(Throwable tx)
+          {
+            errors.add(new PluginThrowable((Plugin)plugin, tx));
```

Comment:
The system println calls should instead be in here. Note: the stacktrace shows the code that caused this to be called, so we can always tell what kind of event triggered the error.

Reply:
Eventually, they will be replaced with Logging calls, once we actually start work on #50.